### PR TITLE
Add `typeable`

### PR DESCRIPTION
**Pull request by zardyh**
wowie, magic classes

### CODE REVIEW

File: src/Syntax/Implicits.hs [hunk 1]
```diff
@@ -284,6 +286,13 @@ matches (TyPromotedCon t) (TyPromotedCon t') = t == t'
 matches TyPromotedCon{} _ = False
 
 matches (TyApp f x) (TyApp f' x') = f `matches` f' && x `matches` x'
+
+matches (TyApp f x) (a :-> b) =
+  matches f (TyApp (TyCon (TgName (T.pack "->") (negate 7))) a) && matches x b
```
Comment (SquidDev) [inline src/Syntax/Implicits.hs]: This seems incredibly fragile. How are we doing this elsewhere?

File: src/Backend/Lua/Builtin.hs [hunk 2]
```diff
@@ -215,5 +215,60 @@ builtins =
...
+                        / 10^(math.floor(math.log(math.abs(ta.fingerprint)) + math.log(math.abs(tb.fingerprint))))
+          else
+            finger = ta.fingerprint + 10^(math.floor(math.log(-ta.fingerprint))) * tb.fingerprint
+          end
+
+          return { { fingerprint = finger, name = ta.name .. " :$ " .. tb.name } , __tag = "TypeRep" }
+        end
+      |] )
+
+-- Type _Typeable_app instance makes a decimal of the concatenation of
```
Comment (SquidDev) [inline src/Backend/Lua/Builtin.hs]: Hrmr, can you explain? Is this a backend issue, is there anything I can do to fix it?

File: src/Syntax/Implicits.hs (hunk 1 above)
Comment (zardyh) [reply src/Syntax/Implicits.hs]: Exactly like this (but using the builtin name)

Comment (zardyh) [reply src/Syntax/Implicits.hs]: (done)

File: src/Control/Monad/Infer.hs [hunk 3]
```diff
@@ -485,6 +488,12 @@ instance Pretty TypeError where
   pretty (NotAClass name) =
     vsep [ "Can not make an instance of" <+> pretty name <+> "because it is not a class" ]
 
+  pretty (DIMalformedHead _) =
+    "Malformed head in" <+> keyword "deriving instance" <+> "clause"
+
+  pretty (DICan'tDerive n _) =
+    "Can't derive an instance of class" <+> stypeSkol (pretty n)
```
Comment (SquidDev) [inline src/Control/Monad/Infer.hs]: Should be "I can't derive, whether you should live or die".

**Pull request merged**